This directory contains data for this project and scripts for obtaining - and cleaning NHGRI data. __prepare_data.sh__ will download the NHGRI GWAS catalog and archive every version in the __catalog_archive__ folder. The most _recent_ version of the gwas catalog is __gwascatalog.txt__.# Data Dictionary<table>	<tr>		<th>			Variable		</th>		<th>			Description		</th>		<th>			Comments		</th>	</tr>	<tr>		<td>			DATE ADDED TO CATALOG		</td>		<td>			Date Added		</td>		<td>			Convert to date format		</td>	</tr>	<tr>		<td>			PUBMEDID		</td>		<td>			PubMed identification number		</td>		<td>-</td>	</tr>	<tr>		<td>			FIRST AUTHOR		</td>		<td>			Last name of first author		</td>		<td>Convert to Factor</td>	</tr>	<tr>		<td>			DATE		</td>		<td>			Publication date (online (epub) date if available)		</td>		<td>			Convert to date format		</td>	</tr>	<tr>		<td>			JOURNAL		</td>		<td>			Abbreviated journal name		</td>		<td>Convert to Factor Format</td>	</tr>	<tr style='color:red;'>		<td>			LINK		</td>		<td>			PubMed URL		</td>		<td>Drop; Standard Format</td>	</tr>	<tr>		<td>			STUDY		</td>		<td>			Title of paper (linked to PubMed abstract)		</td>		<td></td>	</tr>	<tr>		<td>			DISEASE/TRAIT		</td>		<td>			Disease or trait examined in study		</td>		<td>Convert to factor format</td>	</tr>	<tr>		<td>			INITIAL SAMPLE SIZE		</td>		<td>			Sample size for Stage 1 of GWAS		</td>		<td></td>	</tr>	<tr>		<td>			REPLICATION SAMPLE SIZE		</td>		<td>			Sample size for subsequent replication(s)		</td>		<td></td>	</tr>	<tr>		<td>			REGION		</td>		<td>			Cytogenetic region associated with rs number (NCBI)		</td>		<td></td>	</tr>	<tr>		<td>			REPORTED GENE (S)		</td>		<td>			Gene(s) reported by author		</td>		<td></td>	</tr>	<tr>		<td>			MAPPED GENE(S)		</td>		<td>			Gene(s) mapped to the strongest SNP (NCBI). If the SNP is located within a gene, that gene is listed. If the SNP is intergenic, the upstream and downstream genes are listed, separated by a hyphen. Gene info is linked to Entrez Gene.		</td>		<td></td>	</tr>	<tr>		<td>			STRONGEST SNP-RISK ALLELE		</td>		<td>			SNP(s) most strongly associated with trait + risk allele (? for unknown risk allele). May also refer to a haplotype. SNP info is linked to dbSNP.		</td>		<td></td>	</tr>	<tr>		<td>			SNPS		</td>		<td>			Strongest SNP; if a haplotype is reported above, may include more than one rs number (multiple SNPs comprising the haplotype). Asterisk indicates that the published rs number has since been merged with a different rs number (linked to current record in dbSNP).		</td>		<td></td>	</tr>	<tr>		<td>			CONTEXT		</td>		<td>			SNP functional class (NCBI)		</td>		<td></td>	</tr>	<tr>		<td>			RISK ALLELE FREQUENCY		</td>		<td>			Reported risk allele frequency associated with strongest		</td>		<td></td>	</tr>	<tr>		<td>			SNP		</td>		<td></td>	</tr>	<tr>		<td>			P-VALUE		</td>		<td>			Reported p-value for strongest SNP risk allele (linked to dbGaP Association Browser)		</td>		<td></td>	</tr>	<tr>		<td>			P-VALUE (TEXT)		</td>		<td>			Information describing context of p-value (e.g. females, smokers). Note that p-values are rounded to 1 significant digit (for example, a published p-value of 4.8 x 10-7 is rounded to 5 x 10-7).		</td>		<td></td>	</tr>	<tr>		<td>			OR or BETA		</td>		<td>			Reported odds ratio or beta-coefficient associated with strongest SNP risk allele		</td>		<td></td>	</tr>	<tr>		<td>			95% CI (TEXT)		</td>		<td>			Reported 95% confidence interval associated with strongest SNP risk allele		</td>		<td></td>	</tr>	<tr>		<td>			PLATFORM (SNPS PASSING QC)		</td>		<td>			Genotyping platform manufacturer used in Stage 1; also includes notation of pooled DNA study design or imputation of SNPs, where applicable		</td>		<td></td>	</tr>	<tr>		<td>			CNV		</td>		<td>			Study of copy number variation (yes/no)		</td>		<td></td>	</tr></table>